src/overlay/TCPPeer.cpp

```diff
@@ -382,48 +446,10 @@ TCPPeer::drop()
     auto self = static_pointer_cast<TCPPeer>(shared_from_this());
     getApp().getOverlayManager().dropPeer(this);
 
-    // To shutdown, we first queue up our desire to shutdown in the strand,
-    // behind any pending read/write calls. We'll let them issue first.
-    getApp().getClock().getIOService().post([self]() {
-        // Gracefully shut down connection: this pushes a FIN packet into
-        // TCP which, if we wanted to be really polite about, we would wait
-        // for an ACK from by doing repeated reads until we get a 0-read.
-        //
-        // But since we _might_ be dropping a hostile or unresponsive
-        // connection, we're going to just post a close() immediately after,
-        // and hope the kernel does something useful as far as putting
-        // any queued last-gasp ERROR_MSG packet on the wire.
-        //
-        // All of this is voluntary. We can also just close(2) here and
-        // be done with it, but we want to give some chance of telling
-        // peers why we're disconnecting them.
-        asio::error_code ec;
-        self->mSocket->next_layer().shutdown(
-            asio::ip::tcp::socket::shutdown_both, ec);
-        if (ec)
-        {
-            CLOG(ERROR, "Overlay")
-                << "TCPPeer::drop shutdown socket failed: " << ec.message();
-        }
-        self->getApp().getClock().getIOService().post([self]() {
-            // Close fd associated with socket. Socket is already
-            // shut down, but depending on platform (and apparently
-            // whether there was unread data when we issued
-            // shutdown()) this call might push RST onto the wire,
-            // or some other action; in any case it has to be done
-            // to free the OS resources.
-            //
-            // It will also, at this point, cancel any pending asio
-            // read/write handlers, i.e. fire them with an error
-            // code indicating cancellation.
-            asio::error_code ec2;
-            self->mSocket->close(ec2);
-            if (ec2)
-            {
-                CLOG(ERROR, "Overlay")
-                    << "TCPPeer::drop close socket failed: " << ec2.message();
-            }
-        });
-    });
+    // if write queue is not empty, messageSender will take care of shutdown
```

Comment:
> messageSender will take care of shutdown

this is not always the case:
* see my comment in `messageSender` on how to fix the "normal shutdown"
* you also need to keep the `mIdleTimer` in case writes "never" complete. Note that during shutdown we do not want to reset that timer when receiving data - I would suggest looking for all calls to `receivedBytes` and make sure that we stop processing reads in each of those functions with:
```
    if (shouldAbort())
    {
        return;
    }
```

Reply:
Also as a condition, checking if `mWriting` is true is cleaner

Reply:
Also, we may want to add a parameter to `drop` to shutdown immediately (probably the default) as opposed to do this delayed shutdown (in case of timeout like with `Peer::idleTimerExpired` or bad behavior we probably want to stop the connection asap and not wait for flushing).